File: iosApp/iosApp/note_list/NoteListViewModel.swift
//
//  NoteListViewModel.swift
//  iosApp
//
//  Created by Charles Yost on 10/2/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

extension NoteListScreen {
    @MainActor class NoteListViewModel: ObservableObject {
        private var noteDataSource: NoteDataSource? = nil
        
        private let searchNotes = SearchNotes()
        
        private var notes = [Note]()
        @Published private(set)var filteredNotes = [Note]()
        @Published var searchText = "" {
            didSet {
                self.searchNotes.execute(notes: self.notes, query: searchText)
            }
        }
        @Published private(set) var isSearchActive = false
        
        init(noteDataSource: NoteDataSource? = nil) {
            self.noteDataSource = noteDataSource
        }
        
        func loadNotes() {
            noteDataSource?.getAllNotes(completionHandler: { notes, error in
                self.notes = notes ?? []
                self.filteredNotes = self.notes
            })
        }
        
        func deleteNoteById(id: Int64?) {
            if id != nil {
                noteDataSource?.deleteNoteById(id: id!, completionHandler: { error in
                    self.loadNotes()
                })
            }
        }
        
        func toggleIsSearchActive() {
            isSearchActive = !isSearchActive
            if !isSearchActive {
                searchText = ""
            }
        }
        
        func setNoteDataSource(noteDataSource: NoteDataSource) {
            self.noteDataSource = noteDataSource
//            noteDataSource.insertNote(note: Note(id: nil, title: "Note title", content: "Note content", colorHex: 0xFF2355, created: DateTimeUtil().now()),
//                                      completionHandler: { error in
//                
//            })
        }
    }
}

File: androidApp/src/main/java/com/pbussolutions/noteappkmp/android/note_list/NoteListViewModel.kt
package com.pbussolutions.noteappkmp.android.note_list

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pbussolutions.noteappkmp.domain.note.Note
import com.pbussolutions.noteappkmp.domain.note.NoteDataSource
import com.pbussolutions.noteappkmp.domain.note.SearchNotes
import com.pbussolutions.noteappkmp.domain.time.DateTimeUtil
import com.pbussolutions.noteappkmp.presentation.RedOrangeHex
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Created by Charles Yost on 9/28/23.
 * <p>
 * </p>
 */
@HiltViewModel
class NoteListViewModel @Inject constructor(
	private val noteDataSource: NoteDataSource,
	private val savedStateHandle: SavedStateHandle
): ViewModel() {

	private val searchNotes = SearchNotes()
	private val notes = savedStateHandle.getStateFlow("notes", emptyList<Note>())
	private val searchText = savedStateHandle.getStateFlow("searchText", "")
	private val isSearchActive = savedStateHandle.getStateFlow("isSearchActive", false)

	val state = combine(notes, searchText, isSearchActive) { notes, searchText, isSearchActive ->
		NoteListState(
			notes = searchNotes.execute(notes, searchText),
			searchText = searchText,
			isSearchActive = isSearchActive
		)
	}.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), NoteListState())

//	init {
//		viewModelScope.launch {
//			(1 .. 10).forEach {
//				noteDataSource.insertNote(
//					Note(
//						id = null,
//						title = "Note$it",
//						content = "Content$it",
//						colorHex = RedOrangeHex,
//						created = DateTimeUtil.now()
//					)
//				)
//			}
//		}
//	}
	fun loadNotes() {
		viewModelScope.launch {
			savedStateHandle["notes"] = noteDataSource.getAllNotes()
		}
	}

	fun onSearchTextChange(text: String) {
		savedStateHandle["searchText"] = text
	}

	fun onToggleSearch() {
		savedStateHandle["isSearchActive"] = !isSearchActive.value
		if(!isSearchActive.value) {
			savedStateHandle["searchText"] = ""
		}
	}

	fun deleteNoteById(id: Long) {
		viewModelScope.launch {
			noteDataSource.deleteNoteById(id)
			loadNotes()
		}
	}
}